src/linux.rs

```diff
@@ -227,10 +227,122 @@ pub fn tproxy_setup(tproxy_args: &TproxyArgs) -> std::io::Result<TproxyState> {
+        run_command("iptables", args)?;
+
+        // sudo iptables -A FORWARD -i enp0s31f6 -o tun10 -j ACCEPT
+        let args = &[
+            "-A",
+            "FORWARD",
+            "-i",
+            default_interface.as_str(),
+            "-o",
+            tun_name.as_str(),
+            "-j",
+            "ACCEPT",
+        ];
+        run_command("iptables", args)?;
+
+        // sudo iptables -A FORWARD -i tun10 -o enp0s31f6 -m state --state RELATED,ESTABLISHED -j ACCEPT
+        let args = &[
+            "-A",
+            "FORWARD",
+            "-i",
+            tun_name.as_str(),
+            "-o",
+            default_interface.as_str(),
+            "-m",
+            "state",
+            "--state",
+            "RELATED,ESTABLISHED",
+            "-j",
+            "ACCEPT",
+        ];
+        run_command("iptables", args)?;
+
+        // sudo sysctl -w net.ipv4.ip_forward=1
+        run_command("sysctl", &["-w", "net.ipv4.ip_forward=1"])?;
+
+        state.restore_port_forwarding = true;
+        #[cfg(feature = "log")]
+        log::debug!("restore port forwarding: {}", state.restore_port_forwarding);
+
+        let mut restore_gateway_mode = Vec::new();
+
+        // sudo iptables -t nat -D POSTROUTING -o tun10 -j MASQUERADE
+        restore_gateway_mode.push(format!("-t nat -D POSTROUTING -o {}", tun_name));
+
+        // sudo iptables -D FORWARD -i enp0s31f6 -o tun10 -j ACCEPT
+        restore_gateway_mode.push(format!("-D FORWARD -i {} -o {}", default_interface, tun_name));
+
+        // sudo iptables -D FORWARD -i tun10 -o enp0s31f6 -m state --state RELATED,ESTABLISHED -j ACCEPT
+        restore_gateway_mode.push(format!(
+            "-D FORWARD -i {} -o {} -m state --state RELATED,ESTABLISHED",
+            tun_name, default_interface
+        ));
+
+        state.restore_gateway_mode = Some(restore_gateway_mode);
+        #[cfg(feature = "log")]
+        log::debug!("restore gateway mode: {:?}", state.restore_gateway_mode);
+    }
+
+    // check for socket fwmark
+    if let Some(fwmark) = tproxy_args.socket_fwmark {
+        let mark = format!("{}", fwmark);
+        let table = "100";
+
+        // sudo ip rule add fwmark 99 table 2022
+        let args = &["rule", "add", "fwmark", mark.as_str(), "table", table];
+        run_command("ip", args)?;
+
+        // get default interface
+        let (default_ip, default_interface) = get_default_gateway()?;
+        let default_ip = format!("{}", default_ip);
+
+        // sudo ip route flush table 2022
+        let args = &["route", "flush", "table", table];
+        run_command("ip", args)?;
+
+        // sudo ip route add table 100 default via 89.0.142.86 dev tun0
+        let args = &[
+            "route",
+            "add",
+            "table",
+            table,
+            "default",
+            "via",
+            default_ip.as_str(),
+            "dev",
+            default_interface.as_str(),
+        ];
```

Comment:
I'd suggest using the `get_restore_components` (which could be renamed to `get_route_components`) function here and use that instead of manually specifying "via" and "dev" arguments. Not all default routes do have a next L3 hop (PPPoE).

Reply:
I renamed the  `get_restore_components` method to `get_route_components` and used it, Thanks.